File: VB/FillEmptySpaceGridColumn/FillGridView.vb
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Views.Grid

Namespace FillEmptySpaceGridColumn
    Public Class FillGridView
        Inherits GridView

        Private needColumnRecalc As Boolean = False

        Public Sub New(ByVal ownerGrid As GridControl)
            MyBase.New(ownerGrid)
        End Sub
        Public Sub New()
        End Sub

        Protected Overridable Sub RecalculateColumnWidths()
            Dim colToResize As FillGridColumn = Nothing
            Dim totalWidth As Integer = 0
            For i As Integer = 0 To Columns.Count - 1
                If Not Columns(i).Visible Then
                    Continue For
                End If
                If CType(Columns(i), FillGridColumn).FillEmptySpace Then
                    colToResize = CType(Columns(i), FillGridColumn)
                Else
                    totalWidth += Columns(i).Width
                End If
            Next i

            If colToResize IsNot Nothing AndAlso ViewInfo.ViewRects.ColumnPanelWidth > 0 Then
                colToResize.Width = ViewInfo.ViewRects.ColumnPanelWidth - totalWidth
            End If
        End Sub

        Public Overrides Sub LayoutChanged()
            MyBase.LayoutChanged()
            If needColumnRecalc Then
                Return
            End If

            needColumnRecalc = True
            If Not OptionsView.ColumnAutoWidth Then
                RecalculateColumnWidths()
            End If
            needColumnRecalc = False
        End Sub

        Protected Overrides ReadOnly Property ViewName() As String
            Get
                Return "FillGridView"
            End Get
        End Property

        Protected Overrides Function CreateColumnCollection() As GridColumnCollection
            Return New FillGridColumnCollection(Me)
        End Function

    End Class
End Namespace

File: CS/FillEmptySpaceGridColumn/FillGridView.cs
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Views.Grid;

namespace FillEmptySpaceGridColumn
{
	public class FillGridView : GridView
	{
		private bool needColumnRecalc = false;

		public FillGridView(GridControl ownerGrid)
			: base(ownerGrid)
		{
		}
		public FillGridView()
		{
		}

		protected virtual void RecalculateColumnWidths()
		{
			FillGridColumn colToResize = null;
			int totalWidth = 0;
			for ( int i = 0; i < Columns.Count; i++ )
			{
                if (!Columns[i].Visible)
                    continue;
				if ( ((FillGridColumn)Columns[i]).FillEmptySpace )
					colToResize = (FillGridColumn)Columns[i];
				else
					totalWidth += Columns[i].Width;
			}

			if ( colToResize != null && ViewInfo.ViewRects.ColumnPanelWidth > 0 )
				colToResize.Width = ViewInfo.ViewRects.ColumnPanelWidth - totalWidth;
		}

		public override void LayoutChanged()
		{
			base.LayoutChanged();
			if ( needColumnRecalc )
				return;

			needColumnRecalc = true;
			if ( !OptionsView.ColumnAutoWidth )
				RecalculateColumnWidths();
			needColumnRecalc = false;
		}

		protected override string ViewName
		{
			get
			{
				return "FillGridView";
			}
		}

		protected override GridColumnCollection CreateColumnCollection()
		{
			return new FillGridColumnCollection(this);
		}

	}
}